src/data/setoid/partition.lean

```diff
@@ -201,3 +212,131 @@ _ (subtype (@is_partition α)) _ (partial_order.to_preorder _) $ partition.rel_i
 end partition
 
 end setoid
+
+/-- A partition of a type index by another type. -/
+@[nolint has_inhabited_instance]
+structure indexed_partition {ι α : Type*} (s : ι → set α) :=
+(eq_of_mem : ∀ {x i j}, x ∈ s i → x ∈ s j → i = j)
+(some : ι → α)
+(some_mem : ∀ i, some i ∈ s i)
```

Comment:
Do indexed partitions always require the sets to be nonempty? An alternate definition could be something like `function.left_inverse (coe ∘ sigma.snd : (Σ i, s i) → α)`

Reply:
I don't why we would require this from partition and not from indexed partition. There shouldn't be any math difference between those notions. They are two different computer implementations of exactly the same idea.

Reply:
I can think of something contrived like considering naturals equivalent if the sum of their unique prime factors is equal. The obvious index would be the sum of factors, but the definition here would not allow that, as `s 4` is empty

Reply:
I don't want to introduce some weird mathematical innovation here, I only want another implementation of a well-known concept because I have a use case. I think you could introduce your concept when you'll actually need it, with a name clearly explaining this is not the standard thing, like `indexed_wieser_pseudo_partition`.

Reply:
That's fine, I've realized the version I'm thinking of is just `index : α → ι` with setoid `setoid.ker index`, while the version you're after is the same with `function.surjective index`. Perhaps definitional control of the sets is important to you though, which would be a good reason to use the version in this PR.